packages/flutter_tools/lib/src/commands/drive.dart

```diff
@@ -203,6 +207,19 @@ class DriveCommand extends RunCommandBase {
   @override
   bool get cachePubGet => false;
 
+  // Network devices need `publish-port` to be enabled because it requires mDNS.
+  // If the flag wasn't provided as an actual argument and it's a network device,
+  // change it to be enabled.
+  @override
+  bool get disablePortPublication {
+    final ArgResults? localArgResults = argResults;
+    if (isNetworkDevice && localArgResults != null && !localArgResults.wasParsed('publish-port')) {
```

Comment:
This seems like a foot-gun for future contributors since `_checkIfNetworkDevice` must run before this method is called to populate `isNetworkDevice`, but that order isn't enforced.

How about something like this:

flutter_command `disablePortPublication` declaration becomes `Future<bool>`
```dart
  Future<bool> get disablePortPublication async => !boolArgDeprecated('publish-port');
```
In the drive command:
```dart
  @override
  Future<bool> get disablePortPublication async {
    final ArgResults? localArgResults = argResults;
    final Device? targetedDevice = await device;
    final bool isNetworkDevice = targetedDevice is IOSDevice && targetedDevice.interfaceType == IOSDeviceConnectionInterface.network;
    if (isNetworkDevice && localArgResults != null && !localArgResults.wasParsed('publish-port')) {
      _logger.printTrace('Network device is being used. Changing `publish-port` to be enabled.');
      return false;
    }
    return !boolArgDeprecated('publish-port');
  }

  String? get applicationBinaryPath => stringArgDeprecated(FlutterOptions.kUseApplicationBinary);

  Future<Device?> get device => findTargetDevice(includeUnsupportedDevices: applicationBinaryPath == null);
```
And the `device` property is used through the rest of the command wherever `findTargetDevice()` is currently used.

Then in the drive tests fake out network or USB devices (note the additional `DeviceManager: () => fakeDeviceManager` override):
```dart
  testUsingContext('Port publication not disabled for network device', () async {
    final DriveCommand command = DriveCommand(
      fileSystem: fileSystem,
      logger: logger,
      platform: platform,
      signals: signals,
    );

    fileSystem.file('lib/main.dart').createSync(recursive: true);
    fileSystem.file('test_driver/main_test.dart').createSync(recursive: true);
    fileSystem.file('pubspec.yaml').createSync();

    final Device networkDevice = FakeIosDevice()
      ..interfaceType = IOSDeviceConnectionInterface.network;
    fakeDeviceManager.devices = <Device>[networkDevice];

    await expectLater(() => createTestCommandRunner(command).run(<String>[
      'drive',
    ]), throwsToolExit());

    final DebuggingOptions options = await command.createDebuggingOptions(false);
    expect(options.disablePortPublication, false);
  }, overrides: <Type, Generator>{
    Cache: () => Cache.test(processManager: FakeProcessManager.any()),
    FileSystem: () => MemoryFileSystem.test(),
    ProcessManager: () => FakeProcessManager.any(),
    DeviceManager: () => fakeDeviceManager,
  });

...
// ignore: avoid_implementing_value_types
class FakeIosDevice extends Fake implements IOSDevice {
  @override
  IOSDeviceConnectionInterface interfaceType = IOSDeviceConnectionInterface.usb;

  @override
  Future<TargetPlatform> get targetPlatform async => TargetPlatform.ios;
}
```